diff --recursive -u ../clean/dropbear-2016.74/svr-chansession.c ./dropbear-2016.74/svr-chansession.c
--- ../clean/dropbear-2016.74/svr-chansession.c	2016-07-21 11:17:09.000000000 -0400
+++ dropbear-2016.74/svr-chansession.c	2017-09-19 11:01:34.889121503 -0400
@@ -925,10 +925,10 @@
 		if ((setgid(ses.authstate.pw_gid) < 0) ||
 			(initgroups(ses.authstate.pw_name, 
 						ses.authstate.pw_gid) < 0)) {
-			dropbear_exit("Error changing user group");
+			//dropbear_exit("Error changing user group");
 		}
 		if (setuid(ses.authstate.pw_uid) < 0) {
-			dropbear_exit("Error changing user");
+			//dropbear_exit("Error changing user");
 		}
 	} else {
 		/* ... but if the daemon is the same uid as the requested uid, we don't
