--- fbwhiptail-0f14a409735b71c219e0b9b3ee63cdae709ba143/whiptail.orig	2022-11-14 14:12:40.556000000 -0500
+++ fbwhiptail-0f14a409735b71c219e0b9b3ee63cdae709ba143/whiptail	2022-11-14 14:15:03.362000000 -0500
@@ -1,2 +1,5 @@
 #!/bin/sh
+if [ $(tty) == "/dev/hvc0" ]; then
+       die "fbwhiptail menus cannot be drawn serial/non-fb enabled consoles. Run commands manually."
+fi 
 fbwhiptail --notags "$@"
