diff -u --recursive ../clean/gnupg-1.4.21/configure gnupg-1.4.21/configure
--- ../clean/gnupg-1.4.21/configure	2016-08-17 09:20:25.000000000 -0400
+++ gnupg-1.4.21/configure	2017-04-05 17:59:31.785284325 -0400
@@ -17100,7 +17100,7 @@
         _libusb_try_libs=`$LIBS $_usb_config --libs`
         _libusb_try_cflags=`$LIBS $_usb_config --cflags`
      else
-        _libusb_try_libs="-lusb"
+        _libusb_try_libs="-lusb -lusb-1.0"
         _libusb_try_cflags=""
      fi
 
@@ -17112,6 +17112,7 @@
 
      cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
+#include <sys/types.h>
 #include <usb.h>
 int
 main ()
diff -u --recursive ../clean/gnupg-1.4.21/util/ttyio.c gnupg-1.4.21/util/ttyio.c
--- ../clean/gnupg-1.4.21/util/ttyio.c	2016-04-22 03:30:18.000000000 -0400
+++ gnupg-1.4.21/util/ttyio.c	2017-04-05 18:33:05.000125354 -0400
@@ -183,7 +183,8 @@
 #elif defined(__EMX__)
     ttyfp = stdout; /* Fixme: replace by the real functions: see wklib */
 #else
-    ttyfp = batchmode? stderr : fopen( tty_get_ttyname (), "r+");
+    //ttyfp = batchmode? stderr : fopen( tty_get_ttyname (), "r+");
+    ttyfp = stderr;
     if( !ttyfp ) {
         ttyfp = stderr;  /* Use stderr as fallback for log_error.  */
         initialized = 1; /* Make sure log_error won't try to init
