--- a/libnitrokey/log.h	2019-06-24 11:23:44.000000000 +0200
+++ b/libnitrokey/log.h	2020-07-28 14:39:15.856979069 +0200
@@ -100,6 +100,7 @@
 #ifdef NO_LOG
 #define LOG(string, level) while(false){}
 #define LOGD(string) while(false){}
+#define LOGD1(string) while(false){}
 #else
 #define LOG(string, level) nitrokey::log::Log::instance()((string), (level))
 #define LOGD1(string) nitrokey::log::Log::instance()((string), (nitrokey::log::Loglevel::DEBUG_L1))
