diff -u --recursive /home/hudson/build/clean/linux-4.7/arch/x86/boot/compressed/head_64.S ./linux-4.7/arch/x86/boot/compressed/head_64.S
--- /home/hudson/build/clean/linux-4.7/arch/x86/boot/compressed/head_64.S	2016-07-24 15:23:50.000000000 -0400
+++ linux-4.7/arch/x86/boot/compressed/head_64.S	2016-08-05 12:07:11.399854225 -0400
@@ -340,9 +357,15 @@
 1:
 
 	/* Target address to relocate to for decompression */
+#if 0
 	movl	BP_init_size(%rsi), %ebx
 	subl	$_end, %ebx
 	addq	%rbp, %rbx
+#else
+	// coreboot does not populate the init_size boot param,
+	// so we have to fake it with a hard coded value
+	movl    $0x97b000, %ebx
+#endif
 
 	/* Set up the stack */
 	leaq	boot_stack_end(%rbx), %rsp
