diff --git a/config.sh b/config.sh
index 4e321c9..6d9ea32 100644
--- a/config.sh
+++ b/config.sh
@@ -1,13 +1,15 @@
 # ARCH will be auto-detected as the host if not specified
 #ARCH=i486
-#ARCH=x86_64
+ARCH=x86_64
 #ARCH=powerpc
 #ARCH=arm
 #ARCH=microblaze
 #ARCH=mips
 #ARCH=mipsel
 
-CC_BASE_PREFIX=/opt/cross
+if [ -z "$CC_BASE_PREFIX" ]; then
+	CC_BASE_PREFIX=$HOME/heads/install
+fi
 
 # If you use arm, you may need more fine-tuning:
 # arm hardfloat v7
@@ -20,7 +22,10 @@ CC_BASE_PREFIX=/opt/cross
 #GCC_BOOTSTRAP_CONFFLAGS="--with-arch=armv7-a --with-float=softfp"
 #GCC_CONFFLAGS="--with-arch=armv7-a --with-float=softfp"
 
-MAKEFLAGS=-j8
+MAKEFLAGS=-j`nproc`
 
 # Enable this to build the bootstrap gcc (thrown away) without optimization, to reduce build time
 GCC_STAGE1_NOOPT=1
+
+# Build GMP, MPFR and MPC
+GCC_BUILTIN_PREREQS=yes
diff --git a/defs.sh b/defs.sh
index f76a2ac..9184123 100644
--- a/defs.sh
+++ b/defs.sh
@@ -221,19 +221,19 @@ muslfetchextract() {
 gccprereqs() {
     if [ ! -e gcc-$GCC_VERSION/gmp ]
     then
-        fetchextract http://gmplib.org/download/gmp/ gmp-$GMP_VERSION .tar.bz2
+        fetchextract https://gmplib.org/download/gmp/ gmp-$GMP_VERSION .tar.bz2
         mv gmp-$GMP_VERSION gcc-$GCC_VERSION/gmp
     fi
 
     if [ ! -e gcc-$GCC_VERSION/mpfr ]
     then
-        fetchextract http://ftp.gnu.org/gnu/mpfr/ mpfr-$MPFR_VERSION .tar.bz2
+        fetchextract https://ftp.gnu.org/gnu/mpfr/ mpfr-$MPFR_VERSION .tar.bz2
         mv mpfr-$MPFR_VERSION gcc-$GCC_VERSION/mpfr
     fi
 
     if [ ! -e gcc-$GCC_VERSION/mpc ]
     then
-        fetchextract http://www.multiprecision.org/mpc/download/ mpc-$MPC_VERSION .tar.gz
+        fetchextract https://ftp.gnu.org/gnu/mpc/ mpc-$MPC_VERSION .tar.gz
         mv mpc-$MPC_VERSION gcc-$GCC_VERSION/mpc
     fi
 }
