diff --git a/config.sh b/config.sh
index 4e321c9..336c65b 100644
--- a/config.sh
+++ b/config.sh
@@ -1,13 +1,15 @@
 # ARCH will be auto-detected as the host if not specified
 #ARCH=i486
-#ARCH=x86_64
+ARCH=x86_64
 #ARCH=powerpc
 #ARCH=arm
 #ARCH=microblaze
 #ARCH=mips
 #ARCH=mipsel
 
-CC_BASE_PREFIX=/opt/cross
+if [ -z "$CC_BASE_PREFIX" ]; then
+	CC_BASE_PREFIX=$HOME/heads/install
+fi
 
 # If you use arm, you may need more fine-tuning:
 # arm hardfloat v7
@@ -20,7 +22,10 @@ CC_BASE_PREFIX=/opt/cross
 #GCC_BOOTSTRAP_CONFFLAGS="--with-arch=armv7-a --with-float=softfp"
 #GCC_CONFFLAGS="--with-arch=armv7-a --with-float=softfp"
 
-MAKEFLAGS=-j8
+MAKEFLAGS=-j`nproc`
 
 # Enable this to build the bootstrap gcc (thrown away) without optimization, to reduce build time
 GCC_STAGE1_NOOPT=1
+
+# Build GMP, MPFR and MPC
+GCC_BUILTIN_PREREQS=yes
