--- ./configure.orig	2023-05-17 06:53:16.721284360 -0400
+++ ./configure	2023-11-28 13:57:50.012000000 -0500
@@ -16580,7 +16580,7 @@
   version_type=linux # correct to gnu/linux during the next big refactor
   need_lib_prefix=no
   need_version=no
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   if test ia64 = "$host_cpu"; then
     # AIX 5 supports IA64
     library_names_spec='$libname$release$shared_ext$major $libname$release$shared_ext$versuffix $libname$shared_ext'
@@ -16870,16 +16870,16 @@
     ;;
   freebsd3.[01]* | freebsdelf3.[01]*)
     shlibpath_overrides_runpath=yes
-    hardcode_into_libs=yes
+    hardcode_into_libs=no
     ;;
   freebsd3.[2-9]* | freebsdelf3.[2-9]* | \
   freebsd4.[0-5] | freebsdelf4.[0-5] | freebsd4.1.1 | freebsdelf4.1.1)
     shlibpath_overrides_runpath=no
-    hardcode_into_libs=yes
+    hardcode_into_libs=no
     ;;
   *) # from 4.6 on, and DragonFly
     shlibpath_overrides_runpath=yes
-    hardcode_into_libs=yes
+    hardcode_into_libs=no
     ;;
   esac
   ;;
@@ -16894,7 +16894,7 @@
   shlibpath_var=LIBRARY_PATH
   shlibpath_overrides_runpath=no
   sys_lib_dlsearch_path_spec='/boot/home/config/lib /boot/common/lib /boot/system/lib'
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   ;;
 
 hpux9* | hpux10* | hpux11*)
@@ -16906,7 +16906,7 @@
   case $host_cpu in
   ia64*)
     shrext_cmds='.so'
-    hardcode_into_libs=yes
+    hardcode_into_libs=no
     dynamic_linker="$host_os dld.so"
     shlibpath_var=LD_LIBRARY_PATH
     shlibpath_overrides_runpath=yes # Unless +noenvvar is specified.
@@ -16922,7 +16922,7 @@
     ;;
   hppa*64*)
     shrext_cmds='.sl'
-    hardcode_into_libs=yes
+    hardcode_into_libs=no
     dynamic_linker="$host_os dld.sl"
     shlibpath_var=LD_LIBRARY_PATH # How should we handle SHLIB_PATH
     shlibpath_overrides_runpath=yes # Unless +noenvvar is specified.
@@ -16955,7 +16955,7 @@
   dynamic_linker='Interix 3.x ld.so.1 (PE, like ELF)'
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=no
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   ;;
 
 irix5* | irix6* | nonstopux*)
@@ -16992,7 +16992,7 @@
   shlibpath_overrides_runpath=no
   sys_lib_search_path_spec="/usr/lib$libsuff /lib$libsuff /usr/local/lib$libsuff"
   sys_lib_dlsearch_path_spec="/usr/lib$libsuff /lib$libsuff"
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   ;;
 
 # No shared lib support for Linux oldld, aout, or coff.
@@ -17013,7 +17013,7 @@
   # This implies no fast_install, which is unacceptable.
   # Some rework will be needed to allow for fast_install
   # before this can be enabled.
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
 
   dynamic_linker='Android linker'
   # Don't embed -rpath directories since the linker doesn't support them.
@@ -17071,7 +17071,7 @@
   # This implies no fast_install, which is unacceptable.
   # Some rework will be needed to allow for fast_install
   # before this can be enabled.
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
 
   # Add ABI-specific directories to the system library path.
   sys_lib_dlsearch_path_spec="/lib64 /usr/lib64 /lib /usr/lib"
@@ -17111,7 +17111,7 @@
   fi
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=yes
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   ;;
 
 newsos6)
@@ -17129,7 +17129,7 @@
   soname_spec='$libname$release$shared_ext$major'
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=no
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   dynamic_linker='ldqnx.so'
   ;;
 
@@ -17201,7 +17201,7 @@
   soname_spec='$libname$release$shared_ext$major'
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=yes
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   # ldd complains unless libraries are executable
   postinstall_cmds='chmod +x $lib'
   ;;
@@ -17258,7 +17258,7 @@
   soname_spec='$libname$release$shared_ext$major'
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=yes
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   if test yes = "$with_gnu_ld"; then
     sys_lib_search_path_spec='/usr/local/lib /usr/gnu/lib /usr/ccs/lib /usr/lib /lib'
   else
@@ -17280,7 +17280,7 @@
   library_names_spec='$libname$release$shared_ext$versuffix $libname$release$shared_ext$major $libname$shared_ext'
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=no
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   ;;
 
 uts4*)
@@ -20574,7 +20574,7 @@
   version_type=linux # correct to gnu/linux during the next big refactor
   need_lib_prefix=no
   need_version=no
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   if test ia64 = "$host_cpu"; then
     # AIX 5 supports IA64
     library_names_spec='$libname$release$shared_ext$major $libname$release$shared_ext$versuffix $libname$shared_ext'
@@ -20862,16 +20862,16 @@
     ;;
   freebsd3.[01]* | freebsdelf3.[01]*)
     shlibpath_overrides_runpath=yes
-    hardcode_into_libs=yes
+    hardcode_into_libs=no
     ;;
   freebsd3.[2-9]* | freebsdelf3.[2-9]* | \
   freebsd4.[0-5] | freebsdelf4.[0-5] | freebsd4.1.1 | freebsdelf4.1.1)
     shlibpath_overrides_runpath=no
-    hardcode_into_libs=yes
+    hardcode_into_libs=no
     ;;
   *) # from 4.6 on, and DragonFly
     shlibpath_overrides_runpath=yes
-    hardcode_into_libs=yes
+    hardcode_into_libs=no
     ;;
   esac
   ;;
@@ -20886,7 +20886,7 @@
   shlibpath_var=LIBRARY_PATH
   shlibpath_overrides_runpath=no
   sys_lib_dlsearch_path_spec='/boot/home/config/lib /boot/common/lib /boot/system/lib'
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   ;;
 
 hpux9* | hpux10* | hpux11*)
@@ -20898,7 +20898,7 @@
   case $host_cpu in
   ia64*)
     shrext_cmds='.so'
-    hardcode_into_libs=yes
+    hardcode_into_libs=no
     dynamic_linker="$host_os dld.so"
     shlibpath_var=LD_LIBRARY_PATH
     shlibpath_overrides_runpath=yes # Unless +noenvvar is specified.
@@ -20914,7 +20914,7 @@
     ;;
   hppa*64*)
     shrext_cmds='.sl'
-    hardcode_into_libs=yes
+    hardcode_into_libs=no
     dynamic_linker="$host_os dld.sl"
     shlibpath_var=LD_LIBRARY_PATH # How should we handle SHLIB_PATH
     shlibpath_overrides_runpath=yes # Unless +noenvvar is specified.
@@ -20947,7 +20947,7 @@
   dynamic_linker='Interix 3.x ld.so.1 (PE, like ELF)'
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=no
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   ;;
 
 irix5* | irix6* | nonstopux*)
@@ -20984,7 +20984,7 @@
   shlibpath_overrides_runpath=no
   sys_lib_search_path_spec="/usr/lib$libsuff /lib$libsuff /usr/local/lib$libsuff"
   sys_lib_dlsearch_path_spec="/usr/lib$libsuff /lib$libsuff"
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   ;;
 
 # No shared lib support for Linux oldld, aout, or coff.
@@ -21005,7 +21005,7 @@
   # This implies no fast_install, which is unacceptable.
   # Some rework will be needed to allow for fast_install
   # before this can be enabled.
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
 
   dynamic_linker='Android linker'
   # Don't embed -rpath directories since the linker doesn't support them.
@@ -21063,7 +21063,7 @@
   # This implies no fast_install, which is unacceptable.
   # Some rework will be needed to allow for fast_install
   # before this can be enabled.
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
 
   # Add ABI-specific directories to the system library path.
   sys_lib_dlsearch_path_spec="/lib64 /usr/lib64 /lib /usr/lib"
@@ -21103,7 +21103,7 @@
   fi
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=yes
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   ;;
 
 newsos6)
@@ -21121,7 +21121,7 @@
   soname_spec='$libname$release$shared_ext$major'
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=no
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   dynamic_linker='ldqnx.so'
   ;;
 
@@ -21193,7 +21193,7 @@
   soname_spec='$libname$release$shared_ext$major'
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=yes
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   # ldd complains unless libraries are executable
   postinstall_cmds='chmod +x $lib'
   ;;
@@ -21250,7 +21250,7 @@
   soname_spec='$libname$release$shared_ext$major'
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=yes
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   if test yes = "$with_gnu_ld"; then
     sys_lib_search_path_spec='/usr/local/lib /usr/gnu/lib /usr/ccs/lib /usr/lib /lib'
   else
@@ -21272,7 +21272,7 @@
   library_names_spec='$libname$release$shared_ext$versuffix $libname$release$shared_ext$major $libname$shared_ext'
   shlibpath_var=LD_LIBRARY_PATH
   shlibpath_overrides_runpath=no
-  hardcode_into_libs=yes
+  hardcode_into_libs=no
   ;;
 
 uts4*)
