--- ./ichspi.c.orig	2024-08-11 00:40:38.777007492 -0400
+++ ./ichspi.c	2024-08-11 00:46:03.904024919 -0400
@@ -492,6 +492,7 @@
 
 static void prettyprint_ich9_reg_hsfc(uint16_t reg_val, enum ich_chipset ich_gen)
 {
+#if 0
 	msg_pdbg("HSFC: ");
 	pprint_reg16(HSFC, FGO, reg_val, ", ");
 	switch (ich_gen) {
@@ -511,6 +512,7 @@
 	}
 	pprint_reg16(HSFC, FDBC, reg_val, ", ");
 	pprint_reg16(HSFC, SME, reg_val, "\n");
+#endif
 }
 
 static void prettyprint_ich9_reg_ssfs(uint32_t reg_val)
