diff -ur dasharo_cb.org/src/vendorcode/dasharo/options.c dasharo_cb/src/vendorcode/dasharo/options.c
--- dasharo_cb.org/src/vendorcode/dasharo/options.c	2023-12-18 21:57:46.915582991 +0100
+++ dasharo_cb/src/vendorcode/dasharo/options.c	2023-12-19 19:16:13.077019389 +0100
@@ -189,7 +189,12 @@
 
 uint8_t get_sleep_type_option(void)
 {
+
+#if CONFIG(BOARD_NOVACUSTOM_NV4X_ADLP)
+	uint8_t sleep_type = SLEEP_TYPE_OPTION_S3;
+#else
 	uint8_t sleep_type = SLEEP_TYPE_OPTION_S0IX;
+#endif
 
 	if (CONFIG(DRIVERS_EFI_VARIABLE_STORE))
 		read_u8_var("SleepType", &sleep_type);
