diff -ur coreboot-nitrokey.org/src/vendorcode/dasharo/options.c coreboot-nitrokey/src/vendorcode/dasharo/options.c
--- coreboot-nitrokey.org/src/vendorcode/dasharo/options.c	2024-01-22 14:11:59.525612567 +0100
+++ coreboot-nitrokey/src/vendorcode/dasharo/options.c	2024-01-22 14:12:07.535544365 +0100
@@ -133,7 +133,7 @@
 	if (CONFIG(DRIVERS_EFI_VARIABLE_STORE))
 		read_u8_var("MeMode", &var);
 
-	return var;
+	return ME_MODE_DISABLE_HAP;
 }
 
 bool is_smm_bwp_permitted(void)
